NotificationHubs/src/com/windowsazure/messaging/RetryUtil.java

```diff
@@ -72,18 +72,46 @@ public static RetryPolicy getRetryPolicy(RetryOptions options) {
                 }
 
                 if (error instanceof TimeoutException) {
-                	System.out.println("TimeoutException error occurred. Retrying operation. Attempt: " + attempt);
-                	
+                	System.out.println("TimeoutException error occurred. Retrying operation. Attempt: " + attempt); 
                     return retryPolicy.calculateRetryDelay(error, attempt);
-                } else if (error instanceof QuotaExceededException && (((QuotaExceededException) error).getIsTransient())) {
+                } else if (error instanceof QuotaExceededException) {
                 	System.out.println("Retryable error occurred. Retrying operation. Attempt: " + attempt + ". Error: " + error);
-                	
                     return retryPolicy.calculateRetryDelay(error, attempt);
+                } else if (error instanceof NotificationHubsException) {                    	
+                    	int statusCode = ((NotificationHubsException) error).getHttpStatusCode();
+                    	if (statusCode == 408 || statusCode == 500 || statusCode == 503 || statusCode == 504) {
+                    		System.out.println("Retryable error occurred. Retrying operation. Attempt: " + attempt + ". Error: " + error);
+                    		return retryPolicy.calculateRetryDelay(error, attempt);
+                    	}
+                    	
+                    	System.out.println("Error is not a TimeoutException nor is it a retryable exception." + error);
+                        throw Exceptions.propagate(error);
                 } else {
                 	System.out.println("Error is not a TimeoutException nor is it a retryable exception." + error);
                     throw Exceptions.propagate(error);
                 }
             })
             .flatMap(Mono::delay);
     }
+
+    public static Optional<Integer> parseRetryAfter(HttpResponse response)
```

Comment:
I'm not clear on the status of the Apache HTTP library. At least in the Android world, I believe it is treated as deprecated. For that reason, I'm more sensitive than usual about including a type from that library in our public surface area.

At very least, let's make this a package-protected method to defer that decision. Even better would be to just take a String which is the value of the HTTP header.

Reply:
I don't know about Android world, but I took a look on HttpClientManager and we also using Apache HTTP library for almost every action, so tied on this dependency anyway. So, I think we can leave it as is.

Reply:
@marstr i have changed this method to package-protected